### PR TITLE
Bugfix/spk 101. App crash on cancelling photo within the native camera library buttons

### PR DESCRIPTION
**Pull request by AlexMisiulia**

### CODE REVIEW

File: app/src/main/java/com/yoti/mobile/android/social/keys/features/writekey/TakePhotoViewModel.kt [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        }
+
+        val photoPath = capturedPhotoPath ?: return
+        _continueEvent.value = Event(UserPhotoInfo(formName, projectName, photoPath))
+    }
+
+    private fun setTakePhotoError() {
+        _errorEvent.value = Event(R.string.cant_take_photo_error)
+    }
+
```
Comment (AlexMisiulia) [inline app/src/main/java/com/yoti/mobile/android/social/keys/features/writekey/TakePhotoViewModel.kt]: Remove blank lines

File: app/src/main/java/com/yoti/mobile/android/social/keys/features/writekey/TakePhotoViewModel.kt [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+                imageUtils.toBitmapNotNull(currentPhotoPath)
+            }
+
+            capturedPhotoPath = currentPhotoPath
+            _photoLiveData.value = currentPhotoBitmap
+        }
+    }
+
+    fun onContinueClicked() {
+        if (capturedPhotoPath == null) {
```
Comment (AlexMisiulia) [inline app/src/main/java/com/yoti/mobile/android/social/keys/features/writekey/TakePhotoViewModel.kt]: Add the comment

File: app/src/main/java/com/yoti/mobile/android/social/keys/utils/ImageUtils.kt [hunk 3]
```diff
@@ -96,49 +81,17 @@ class ImageUtils @Inject constructor() {
...
-        img.recycle()
-        return rotatedImg
-    }
+    fun toBitmap(photoBytes: ByteArray): Bitmap? = BitmapFactory.decodeByteArray(photoBytes, 0, photoBytes.size)
 
     fun toByteArray(imagePath: String, maxWidth: Int, maxHeight: Int, quality: Int = 100): ByteArray {
         val bitmap = toBitmap(imagePath, maxWidth, maxHeight)
         val outputStream = ByteArrayOutputStream()
-        bitmap.compress(Bitmap.CompressFormat.JPEG, quality, outputStream)
+        bitmap!!.compress(Bitmap.CompressFormat.JPEG, quality, outputStream)
```
Comment (AlexMisiulia) [inline app/src/main/java/com/yoti/mobile/android/social/keys/utils/ImageUtils.kt]: Replace`!!` with toBitmapNotNull call

File: app/src/test/java/com/yoti/mobile/android/social/keys/features/writekey/TakePhotoViewModelTest.kt [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+
+        val startCameraValue = viewModel.startCameraEvent.value?.peekContent()
+        assertEquals(null, startCameraValue)
+    }
+
+    @Test
+    fun `show error when receive camera result and photo path is null`() = runBlocking {
+        // arrange
+        viewModel.init(FORM_NAME, PROJECT_NAME)
+        whenever(cameraInteractor.createFileForPhotoAsync()).thenReturn(null)
```
Comment (AlexMisiulia) [inline app/src/test/java/com/yoti/mobile/android/social/keys/features/writekey/TakePhotoViewModelTest.kt]: Useless mock - remove it

File: app/src/test/java/com/yoti/mobile/android/social/keys/features/writekey/TakePhotoViewModelTest.kt [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+        // assert
+        val errorValue = viewModel.errorEvent.value?.peekContent()
+        assertEquals(R.string.cant_take_photo_error, errorValue)
+    }
+
+    @Test
+    fun `display image when receive camera result`() = runBlocking {
+        // arrange
+        viewModel.init(FORM_NAME, PROJECT_NAME)
+        val file = File("file")
```
Comment (AlexMisiulia) [inline app/src/test/java/com/yoti/mobile/android/social/keys/features/writekey/TakePhotoViewModelTest.kt]: Add blank lines for readability

File: app/src/main/java/com/yoti/mobile/android/social/keys/features/writekey/TakePhotoViewModel.kt (hunk 1 above)
Comment (AlexMisiulia) [reply app/src/main/java/com/yoti/mobile/android/social/keys/features/writekey/TakePhotoViewModel.kt]: Done

File: app/src/main/java/com/yoti/mobile/android/social/keys/features/writekey/TakePhotoViewModel.kt (hunk 2 above)
Comment (AlexMisiulia) [reply app/src/main/java/com/yoti/mobile/android/social/keys/features/writekey/TakePhotoViewModel.kt]: Done

File: app/src/main/java/com/yoti/mobile/android/social/keys/utils/ImageUtils.kt (hunk 3 above)
Comment (AlexMisiulia) [reply app/src/main/java/com/yoti/mobile/android/social/keys/utils/ImageUtils.kt]: Done

File: app/src/test/java/com/yoti/mobile/android/social/keys/features/writekey/TakePhotoViewModelTest.kt (hunk 4 above)
Comment (AlexMisiulia) [reply app/src/test/java/com/yoti/mobile/android/social/keys/features/writekey/TakePhotoViewModelTest.kt]: Done

File: app/src/test/java/com/yoti/mobile/android/social/keys/features/writekey/TakePhotoViewModelTest.kt (hunk 5 above)
Comment (AlexMisiulia) [reply app/src/test/java/com/yoti/mobile/android/social/keys/features/writekey/TakePhotoViewModelTest.kt]: Done

File: app/src/main/java/com/yoti/mobile/android/social/keys/utils/CameraHelper.kt [hunk 6]
```diff
@@ -1,85 +1,17 @@
...
-
-        return try {
-            imageUtils.convertToBitmap(context, photoPath)
-        } catch (throwable: Throwable) {
-            null
-        }
-    }
-}
\ No newline at end of file
+fun getStartCameraIntent(context: Context, photoFile: File): Intent {
```
Comment (cquemin) [inline app/src/main/java/com/yoti/mobile/android/social/keys/utils/CameraHelper.kt]: This method can be moved in the class where it is used.

File: app/src/main/java/com/yoti/mobile/android/social/keys/utils/ImageUtils.kt [hunk 7]
```diff
@@ -69,18 +63,9 @@ class ImageUtils @Inject constructor() {
...
-    fun convertToBitmap(context: Context, imagePath: String): Bitmap {
-
-        val imageUri = Uri.fromFile(File(imagePath))
-        var img: Bitmap = toBitmap(imagePath)
-
-        img = rotateImageIfRequired(context, img, imageUri)
-        return img
-    }
-
     fun toBitmap(
```
Comment (cquemin) [inline app/src/main/java/com/yoti/mobile/android/social/keys/utils/ImageUtils.kt]: Those function should be `suspend` function

File: app/src/main/java/com/yoti/mobile/android/social/keys/utils/ImageUtils.kt [hunk 8]
```diff
@@ -96,47 +81,15 @@ class ImageUtils @Inject constructor() {
...
-
-        val orientation = ei.getAttributeInt(ExifInterface.TAG_ORIENTATION, ExifInterface.ORIENTATION_NORMAL)
-
-        when (orientation) {
-            ExifInterface.ORIENTATION_ROTATE_90 -> return rotateImage(bitmap, 90f)
-            ExifInterface.ORIENTATION_ROTATE_180 -> return rotateImage(bitmap, 180f)
-            ExifInterface.ORIENTATION_ROTATE_270 -> return rotateImage(bitmap, 270f)
-            else -> return bitmap
-        }
+    fun toBitmapNotNull(imagePath: String, maxWidth: Int = MAX_IMAGE_WIDTH, maxHeight: Int = MAX_IMAGE_HEIGHT): Bitmap {
```
Comment (cquemin) [inline app/src/main/java/com/yoti/mobile/android/social/keys/utils/ImageUtils.kt]: `suspend` here too

File: app/src/main/java/com/yoti/mobile/android/social/keys/utils/ImageUtils.kt [hunk 9]
```diff
@@ -96,47 +81,15 @@ class ImageUtils @Inject constructor() {
...
     }
 
-    private fun rotateImage(img: Bitmap, degree: Float): Bitmap {
-        val matrix = Matrix()
-        matrix.postRotate(degree)
-        val rotatedImg = Bitmap.createBitmap(img, 0, 0, img.width, img.height, matrix, true)
-        img.recycle()
-        return rotatedImg
-    }
+    fun toBitmap(photoBytes: ByteArray): Bitmap? = BitmapFactory.decodeByteArray(photoBytes, 0, photoBytes.size)
```
Comment (cquemin) [inline app/src/main/java/com/yoti/mobile/android/social/keys/utils/ImageUtils.kt]: `suspend` here too

File: app/src/main/java/com/yoti/mobile/android/social/keys/utils/CameraHelper.kt (hunk 6 above)
Comment (AlexMisiulia) [reply app/src/main/java/com/yoti/mobile/android/social/keys/utils/CameraHelper.kt]: Done

File: app/src/main/java/com/yoti/mobile/android/social/keys/utils/ImageUtils.kt (hunk 7 above)
Comment (AlexMisiulia) [reply app/src/main/java/com/yoti/mobile/android/social/keys/utils/ImageUtils.kt]: Done

File: app/src/main/java/com/yoti/mobile/android/social/keys/utils/ImageUtils.kt (hunk 8 above)
Comment (AlexMisiulia) [reply app/src/main/java/com/yoti/mobile/android/social/keys/utils/ImageUtils.kt]: Done

File: app/src/main/java/com/yoti/mobile/android/social/keys/utils/ImageUtils.kt (hunk 9 above)
Comment (AlexMisiulia) [reply app/src/main/java/com/yoti/mobile/android/social/keys/utils/ImageUtils.kt]: Done

**Pull request merged**